sdk/core/src/main/java/com/klaviyo/coresdk/networking/KlaviyoApiClient.kt

```diff
@@ -68,55 +71,81 @@ internal object KlaviyoApiClient : ApiClient {
             Registry.dataStore.store(request.uuid, request.toJson())
         }
 
-        Registry.dataStore.store(
-            QUEUE_KEY,
-            JSONArray(apiQueue.map { it.uuid }).toString()
-        )
+        persistQueue()
     }
 
     /**
-     * Awaken queued requests from persistent store
+     * Gets the current size of the API queue
+     *
+     * @return number of requests in the batch queue
+     */
+    fun getQueueSize(): Int = apiQueue.size
+
+    /**
+     * Reset the in-memory queue to the queue from data store
      */
     fun restoreQueue() {
-        // TODO JSON type safety checks
+        while (apiQueue.isNotEmpty()) {
+            apiQueue.remove()
+        }
+
         Registry.dataStore.fetch(QUEUE_KEY)?.let {
-            val queue = JSONArray(it)
-            Array(queue.length()) { i -> queue.optString(i) }
-        }?.map { uuid ->
+            try {
+                val queue = JSONArray(it)
+                Array(queue.length()) { i -> queue.optString(i) }
+            } catch (exception: JSONException) {
+                emptyArray<String>()
+            }
+        }?.forEach { uuid ->
             Registry.dataStore.fetch(uuid)?.let { json ->
-                val jsonObject = JSONObject(json)
-                val request = KlaviyoApiRequest.fromJson(jsonObject)
-                enqueueRequest(request)
+                try {
+                    val request = KlaviyoApiRequest.fromJson(JSONObject(json))
+                    if (!apiQueue.contains(request)) {
+                        apiQueue.offer(request)
```

Comment:
Or is it that we always keep the request in store till send completes, rather than when placed back on the queue?

Reply:
If the latter, it would explain why we are doing a `persistQueue` call below I guess

Reply:
Basically the latter. The reason for calling `persistQueue` here is just in case the process of restoring the queue into memory causes any change, such as parsing errors.

Reply:
I figured this way we're far less likely to lose API requests due to unlucky timing if for example the app quit/crashed and we didn't get the opportunity to write our in memory queue into persistent store before shutdown.